_build/default/bin/dune.exe: Permission denied (ecryptfs, Mint)

<!-- Thank you for filing an issue to help us improve Dune! -->
## Expected Behavior

I can upgrade Dune.

## Actual Behavior

I get a permission denied error.

## Reproduction

I currently have Dune 3.7.1 installed. When I ran `opam upgrade`, Dune 3.9.1 cannot build due to a permission error.

```
[ERROR] The compilation of dune.3.9.1 failed at "./_boot/dune.exe build dune.install --release --profile dune-bootstrap -j 15".

#=== ERROR while compiling dune.3.9.1 =========================================#
# context     2.1.4 | linux/x86_64 | ocaml-base-compiler.5.0.0 | https://opam.ocaml.org#7109cb08
# path        ~/.opam/5.0.0/.opam-switch/build/dune.3.9.1
# command     ~/.opam/opam-init/hooks/sandbox.sh build ./_boot/dune.exe build dune.install --release --profile dune-bootstrap -j 15
# exit-code   1
# env-file    ~/.opam/log/dune-208893-44c9eb.env
# output-file ~/.opam/log/dune-208893-44c9eb.out
### output ###
# Error: _build/default/bin/dune.exe: Permission denied
# -> required by _build/default/bin/dune.exe
# -> required by _build/install/default/bin/dune
# -> required by _build/default/dune.install
```

When I try to upgrade to 3.8.0, 3.8.1, or 3.8.2, I get the error https://github.com/ocaml/dune/issues/8041. When I try to upgrade to 3.8.3, 3.9.0, or 3.9.1, I get this current error. https://github.com/ocaml/dune/pull/8049, which fixes the sendfile error and was included in 3.8.3, is probably the cause of the current error.

My Linux kernel version is `5.15.0-76-generic`, I am running Linux Mint 21.1.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -21,6 +21,9 @@ Unreleased
 - Improve `dune describe external-lib-deps` by adding the internal dependencies
   for more information. (#7478, @moyodiallo)
 
+- Fix permission errors when `sendfile` is not available (#8234, fixes #8120,
+  @emillon)
+
 3.9.1 (2023-07-06)
 ------------------
 

--- a/otherlibs/stdune/src/io.ml
+++ b/otherlibs/stdune/src/io.ml
@@ -101,7 +101,7 @@ module Copyfile = struct
     | Linux -> `Sendfile
     | Windows | Other -> `Nothing
 
-  let sendfile =
+  let sendfile_with_fallback =
     let setup_copy ?(chmod = Fun.id) ~src ~dst () =
       match Unix.openfile src [ O_RDONLY ] 0 with
       | exception Unix.Unix_error (Unix.ENOENT, _, _) -> Error `Src_missing
@@ -147,7 +147,12 @@ module Copyfile = struct
         raise (Sys_error message)
       | Ok (src, dst, src_size) ->
         Exn.protect
-          ~f:(fun () -> sendfile ~src ~dst src_size)
+          ~f:(fun () ->
+            try sendfile ~src ~dst src_size
+            with Unix.Unix_error (EINVAL, "sendfile", _) ->
+              let ic = Unix.in_channel_of_descr src in
+              let oc = Unix.out_channel_of_descr dst in
+              copy_channels ic oc)
           ~finally:(fun () ->
             Unix.close src;
             Unix.close dst)
@@ -175,11 +180,6 @@ module Copyfile = struct
   let copy_file_portable ?chmod ~src ~dst () =
     Exn.protectx (setup_copy ?chmod ~src ~dst ()) ~finally:close_both
       ~f:(fun (ic, oc) -> copy_channels ic oc)
-
-  let sendfile_with_fallback ?chmod ~src ~dst () =
-    try sendfile ?chmod ~src ~dst ()
-    with Unix.Unix_error (EINVAL, "sendfile", _) ->
-      copy_file_portable ?chmod ~src ~dst ()
 
   let copy_file_best =
     match available with

--- a/test/blackbox-tests/test-cases/github8041.t
+++ b/test/blackbox-tests/test-cases/github8041.t
@@ -17,12 +17,6 @@
 If sendfile fails, we should fallback to the portable implementation.
 
   $ strace -e inject=sendfile:error=EINVAL -o /dev/null dune build @install
-  Error: _build/default/data2.txt: Permission denied
-  -> required by _build/default/data2.txt
-  -> required by _build/install/default/share/p/data2.txt
-  -> required by _build/default/p.install
-  -> required by alias install
-  [1]
 
 #8210: data2.txt is copied from readonly-file data.txt (#3092), so it needs to
 be adequately unlinked before starting the fallback.

